janusgraph-backend-testutils/src/main/java/org/janusgraph/testutil/JanusGraphAssert.java

```diff
@@ -81,4 +91,57 @@ private static boolean isEmpty(Object obj) {
         throw new IllegalArgumentException("Cannot determine size of: " + obj);
     }
 
+    public static void assertNumStep(int expectedResults, int expectedSteps, GraphTraversal traversal, Class<? extends Step>... expectedStepTypes) {
+        int num = 0;
+        while (traversal.hasNext()) {
+            traversal.next();
+            num++;
+        }
+
+        assertEquals(expectedResults, num);
+
+        //Verify that steps line up with what is expected after JanusGraph's optimizations are applied
+        List<Step> steps = traversal.asAdmin().getSteps();
+        Set<Class<? extends Step>> expSteps = Sets.newHashSet(expectedStepTypes);
+        int numSteps = 0;
+        for (Step s : steps) {
+            if (s.getClass().equals(GraphStep.class) || s.getClass().equals(StartStep.class)) continue;
+
+            if (expSteps.contains(s.getClass())) {
+                numSteps++;
+            }
+        }
+        assertEquals(expectedSteps, numSteps);
+    }
+
+    public static void assertOptimization(Traversal<?, ?> expectedTraversal, Traversal<?, ?> originalTraversal,
+                                          TraversalStrategy... optimizationStrategies) {
+        final TraversalStrategies optimizations = new DefaultTraversalStrategies();
+        for (final TraversalStrategy<?> strategy : optimizationStrategies) {
+            optimizations.addStrategies(strategy);
+        }
+
+        originalTraversal.asAdmin().setStrategies(optimizations);
+        originalTraversal.asAdmin().applyStrategies();
+
+        assertEquals(expectedTraversal.asAdmin().getSteps().toString(),
+            originalTraversal.asAdmin().getSteps().toString());
```

Comment:
Is there any way to compare steps other than toString()?

Reply:
I didn't find one. If you (or anyone) knows how to do it better, please let me know :)

Reply:
What about simply doing `expectedTraversal.asAdmin().getSteps().equals(originalTraversal.asAdmin().getSteps())`?

Reply:
IIRC there was a problem that equality on steps does not work as expected because of some == / equals discrepancies. I can look at that again and then reply back